security/cmd/node_agent/management/server.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// New creates an NodeAgent server.
+func New(cfg *na.Config) (*Server, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("unablet to create when config is nil")
+	}
+	env := na.DetermineEnv(cfg)
+	pc, err := platform.NewClient(env, cfg.RootCertFile, cfg.KeyFile,
+		cfg.CertChainFile, cfg.IstioCAAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init nodeagent due to pc init %v", err)
+	}
+	scfg := workload.NewSecretFileServerConfig(cfg.CertChainFile, cfg.KeyFile)
+	ss, err := workload.NewSecretServer(scfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to init nodeagent due to secret server %v", err)
+	}
+	return &Server{
+		done:       make(chan bool, 1),
+		handlerMap: make(map[string]handler.WorkloadHandler),
+		cAClient:   &cagrpc.CAGrpcClientImpl{},
+		config:     cfg,
+		pc:         pc,
+		ss:         ss,
+	}, nil
+}
+
+// startLoop loops for the VM mode, i.e not running on k8s.
+// TODO(inclfy): copy paste from nodeagent.go
+func (s *Server) startLoop() { //nolint
+}
+
+// Stop terminates the UDS.
+func (s *Server) Stop() {
+	s.done <- true
+}
+
+// WaitDone waits for a response back from Workloadhandler
+func (s *Server) WaitDone() {
+	<-s.done
+}
+
+// Serve opens the UDS channel and starts to serve NodeAgentService on that uds channel.
+func (s *Server) Serve(path string) {
+	grpcServer := grpc.NewServer()
+	pb.RegisterNodeAgentServiceServer(grpcServer, s)
+
+	var lis net.Listener
+	var err error
+	_, e := os.Stat(path)
+	if e == nil {
+		e := os.RemoveAll(path)
+		if e != nil {
+			log.Errorf("failed to %s with error %v", path, err)
+		}
+	}
+	lis, err = net.Listen("unix", path)
+	if err != nil {
+		log.Errorf("failed to %v", err)
+	}
+
+	go func(ln net.Listener, s *Server) {
+		<-s.done
+		_ = ln.Close()
+		s.CloseAllWlds()
+	}(lis, s)
+
+	_ = grpcServer.Serve(lis)
+}
+
+// WorkloadAdded defines the server side action when a workload is added.
+func (s *Server) WorkloadAdded(ctx context.Context, request *pb.WorkloadInfo) (*pb.NodeAgentMgmtResponse, error) {
+	uid := request.Attrs.Uid
+	sa := request.Attrs.Serviceaccount
+	log.Infof("workload uid = %v, service account = %v added", uid, sa)
+	if _, ok := s.handlerMap[uid]; ok {
+		status := &rpc.Status{Code: int32(rpc.ALREADY_EXISTS), Message: "Already exists"}
+		log.Infof("workload %v already exists", request)
+		return &pb.NodeAgentMgmtResponse{Status: status}, nil
+	}
+
+	// Sends CSR request to CA.
+	// TODO(inclfy): extract the SPIFFE formatting out into somewhere else.
+	id := fmt.Sprintf("spiffe://cluster.local/ns/%s/sa/%s", request.Attrs.Namespace, sa)
+	priv, csrReq, err := s.createRequest(id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create csr request for service %v %v", sa, err)
+	}
+	resp, err := s.cAClient.SendCSR(csrReq, s.pc, s.config.IstioCAAddress)
+	if err != nil {
+		return nil, fmt.Errorf("csr request failed for service %v %v", sa, err)
+	}
+
+	if err := s.ss.SetServiceIdentityPrivateKey(priv); err != nil {
```

Comment:
We will need to merge SetServiceIdentityPrivateKey() and SetServiceIdentityCert() on SecretServer to avoid inconsistency. But we can do that in another PR :)

Reply:
make sens, I logged https://github.com/istio/istio/issues/4406.